pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go

```diff
@@ -0,0 +1,166 @@
+)
+
+// NewAppManifestsDispatcher creates an AppManifestsDispatcher.
+// A K8s client and an ApplicationRevision are required while the resource tracker is optional.
+// If the resource tracker is not provided, the dispatcher will skip GC automatically.
+func NewAppManifestsDispatcher(c client.Client, appRev *v1beta1.ApplicationRevision, latestRT *v1beta1.ResourceTracker) *AppManifestsDispatcher {
+	return &AppManifestsDispatcher{
+		c:          c,
+		applicator: apply.NewAPIApplicator(c),
+		appRev:     appRev.DeepCopy(),
+		oldRT:      latestRT.DeepCopy(),
+		gcHandler:  NewGCHandler(c, appRev.Namespace),
+	}
+}
+
+// AppManifestsDispatcher dispatches application manifests into K8s
+type AppManifestsDispatcher struct {
+	c          client.Client
+	applicator apply.Applicator
+
+	appRev    *v1beta1.ApplicationRevision
+	oldRT     *v1beta1.ResourceTracker
+	skipGC    bool
+	gcHandler GarbageCollector
+
+	appRevName string
+	namespace  string
+	newRTName  string
+	newRT      *v1beta1.ResourceTracker
+}
+
+// SkipGarbageCollect makes dispatcher to skip GC after applying resources
+// It's helpful in a rollout scenario where new revision is going to create a new workload while the old one should not
+// be deleted before rollout is termined.
+func (a *AppManifestsDispatcher) SkipGarbageCollect() {
+	a.skipGC = true
+}
+
+// Dispatch apply manifests into k8s and return a resource tracker recording applied resources
+// If the latest resource tracker is offered to dispatcher, it will do GC after applying.
+func (a *AppManifestsDispatcher) Dispatch(ctx context.Context, manifests []*unstructured.Unstructured) (*v1beta1.ResourceTracker, error) {
+	a.complete()
+	if err := a.createResourceTracker(ctx, manifests); err != nil {
+		return nil, err
+	}
+	if err := a.applyManifests(ctx, manifests); err != nil {
+		return nil, err
+	}
+	if !a.skipGC && a.oldRT != nil && a.oldRT.Name != a.newRTName {
+		if err := a.gcHandler.GarbageCollect(ctx, a.oldRT, a.newRT); err != nil {
+			return nil, errors.WithMessagef(err, "cannot do GC based on resource trackers %q and %q", a.oldRT.Name, a.newRTName)
+		}
+	}
+	return a.newRT.DeepCopy(), nil
+}
+
+func (a *AppManifestsDispatcher) complete() {
+	a.appRevName = a.appRev.Name
+	a.namespace = a.appRev.Namespace
+	a.newRTName = constructResourceTrackerName(a.appRevName, a.namespace)
+}
+
+func (a *AppManifestsDispatcher) createResourceTracker(ctx context.Context, manifests []*unstructured.Unstructured) error {
+	rt := &v1beta1.ResourceTracker{}
+	err := a.c.Get(ctx, client.ObjectKey{Name: a.newRTName}, rt)
+	if err == nil {
+		// already exists
+		// no need to update, because the manifests from the same application revision never change
+		a.newRT = rt
+		return nil
+	}
+	if kerrors.IsNotFound(err) {
+		rt.SetName(a.newRTName)
+		// ApplicationRevision must be controlled by an Application
+		ownerRefOfApp := metav1.GetControllerOf(a.appRev)
+		rt.SetOwnerReferences([]metav1.OwnerReference{*ownerRefOfApp})
```

Comment:
Why application should own rt? As far as I can see, a namespaced scope object (app) cannot own a cluster scope object(resourceTracker) in k8s 1.20

Reply:
This is not used for k8s built-in GC. I'm considering let RT controller take responsibility for GC RT while an application or AppRevision is deleted.

Reply:
I think it's a little bit hack. And I'm not sure k8s whether  truly allow this. how about set some labels in rt to do this.

Reply:
Agreed. If we can make sure the name of resource tracker ALWAYS follows the rule, `name = appRevName+namespace`, we can use its name to locate an AppRevision directly.